include/evaluators/blas2_tree_evaluator.hpp

```diff
@@ -0,0 +1,482 @@
+        r2(Evaluator<RHS2, Device>(expr.r2)) {}
+
+  size_t getSize() { return r1.getSizeR(); }
+  cont_type data() { return r1.data(); }
+
+  bool eval_subexpr_if_needed(cont_type *cont, Device &dev) {
+    r1.eval_subexpr_if_needed(NULL, dev);
+    r2.eval_subexpr_if_needed(NULL, dev);
+    return true;
+  }
+
+  value_type eval(size_t i) {
+    auto dim = r2.getSize();
+    auto valWG =
+        functor_traits<iniAddOp1_struct, value_type, Device>::eval(r2.eval(0));
+    for (size_t j = 0; j < dim; j++) {
+      valWG += r1.eval(i, j) * r2.eval(j);
+    }
+    return valWG;
+  }
+
+  value_type eval(cl::sycl::nd_item<1> ndItem) {
+    return eval(ndItem.get_global(0));
+  }
+
+#if BLAS_EXPERIMENTAL
+  template <typename sharedT>
+  value_type eval(sharedT scratch, cl::sycl::nd_item<1> ndItem) {
+    size_t Pieces = 2;
+
+    size_t localid = ndItem.get_local(0);
+    size_t localSz = ndItem.get_local_range(0);
+    size_t groupid = ndItem.get_group(0);
+    size_t groupSz = ndItem.get_num_groups(0);
+    size_t globalid = ndItem.get_global(0);
+    size_t globalSz = ndItem.get_global_range(0);
+
+    size_t dimR = r1.getSizeR();
+    size_t dimC = r1.getSizeC();
+
+    size_t blqSz = groupSz;  // number of workgroups
+    // row blq id of the current workgroup
+    size_t blqidR = (groupid + (Pieces * blqSz) - 1) / (Pieces * blqSz);
+    size_t blqidC =
+        groupid % (Pieces * blqSz);  // 1st col id of the current workgroup
+
+    // number of columns per each workgroup
+    size_t colSz = (dimC < (Pieces * localSz)) ? dimC : Pieces * localSz;
+    // number of rows per each thread
+    size_t rowSz = (dimR + blqidR - 1) / blqidR;
+
+    size_t colid = blqidC * colSz + localid;  // colid of the current thread
+    size_t rowid = blqidR * rowSz;            // first row of the current thread
+
+    value_type val;
+#if BLAS_EXPERIMENTAL
+    // Local computations
+    while (rowid < dimR) {
+      auto val = functor_traits<iniAddOp1_struct, value_type, Device>::eval(
+          r2.eval(0));
+      for (size_t j = colid; j < dimC; j += colSz) {
+        val += r1.eval(rowid, j) * r2.eval(j);
+      }
+      scratch[localid] = val;
+      // This barrier is mandatory to be sure the data is on the shared memory
+      ndItem.barrier(cl::sycl::access::fence_space::local_space);
+      // Reduction inside the block
+      for (size_t offset = nThr >> 1; offset > 0; offset >>= 1) {
+        if ((rowid < dimR) && (colid < offset)) {
+          scratch[localid] += scratch[localid + offset];
+        }
+        // This barrier is mandatory to be sure the data are on the shared
+        // memory
+        ndItem.barrier(cl::sycl::access::fence_space::local_space);
+      }
+      // The result is stored in lhs
+      if ((rowid < dimR) && (colid == 0)) {
+        l.eval(rowid, blqidC) = scl * scratch[localid] + r3.eval(rowid);
+      }
+      rowid += rowSz;
+    }
+#endif  // BLAS_EXPERIMENTAL
+    return val;
+  }
+#endif  // BLAS_EXPERIMENTAL
```

Comment:
I am not sure, it was there before. I will try to figure out if this code is redundant,